circuit_knitting/cutting/cut_finding/cut_optimization.py

```diff
@@ -0,0 +1,326 @@
+from .search_space_generator import (
+    get_action_subset,
+    SearchFunctions,
+    SearchSpaceGenerator,
+)
+from .disjoint_subcircuits_state import DisjointSubcircuitsState
+from .circuit_interface import SimpleGateList, CircuitElement, Sequence
+from .optimization_settings import OptimizationSettings
+from .quantum_device_constraints import DeviceConstraints
+
+
+@dataclass
+class CutOptimizationFuncArgs:
+    """Collect arguments for passing to the search-space generating methods in :class:`CutOptimization`."""
+
+    entangling_gates: Sequence[Sequence[int | CircuitElement | None | list]] | None = (
+        None
+    )
+    search_actions: ActionNames | None = None
+    max_gamma: float | int | None = None
+    qpu_width: int | None = None
+
+
+def cut_optimization_cost_func(
+    state: DisjointSubcircuitsState, func_args: CutOptimizationFuncArgs
+) -> tuple[int | float, int]:
+    """Return the cost function.
+
+    The particular cost function chosen here aims to minimize the gamma
+    while also (secondarily) giving preference to circuit partitionings
+    that balance the sizes of the resulting partitions, by minimizing the
+    maximum width across subcircuits.
+    """
+    # pylint: disable=unused-argument
+    return (state.lower_bound_gamma(), state.get_max_width())
+
+
+def cut_optimization_upper_bound_cost_func(
+    goal_state, func_args: CutOptimizationFuncArgs
+) -> tuple[int | float, int | float]:
+    """Return the gamma upper bound."""
+    # pylint: disable=unused-argument
+    return (goal_state.upper_bound_gamma(), np.inf)
+
+
+def cut_optimization_min_cost_bound_func(
+    func_args: CutOptimizationFuncArgs,
+) -> tuple[int | float, int | float] | None:
+    """Return an a priori min-cost bound defined in the optimization settings."""
+    if func_args.max_gamma is None:  # pragma: no cover
+        return None
+
+    return (func_args.max_gamma, np.inf)
+
+
+def cut_optimization_next_state_func(
+    state: DisjointSubcircuitsState, func_args: CutOptimizationFuncArgs
+) -> list[DisjointSubcircuitsState]:
+    """Generate a list of next states from the input state."""
+    # Get the entangling gate spec that is to be processed next based
+    # on the search level of the input state
+    assert func_args.entangling_gates is not None
+    assert func_args.search_actions is not None
+
+    gate_spec = func_args.entangling_gates[state.get_search_level()]
+
+    # Determine which search actions can be performed, taking into
+    # account any user-specified constraints that might have been
+    # placed on how the current entangling gate is to be handled
+    # in the search
+    gate = gate_spec[1]
+    gate = cast(CircuitElement, gate)
+    if len(gate.qubits) == 2:
+        action_list = func_args.search_actions.get_group("TwoQubitGates")
+    else:
+        raise ValueError(
+            "In the current version, only the cutting of two qubit gates is supported."
```

Comment:
Currently, if I pass in a circuit with only 3-qubit CCXGates, this will throw an error because it will try to cut those gates. I think it'd be better if the optimizer just didn't consider these gates valid for cutting and ignored them. I'd rather throw a more general error if the optimizer can't find a valid cut scheme only from cutting 2-qubit gates and wires.

Reply:
> I think it'd be better if the optimizer just didn't consider these gates valid for cutting and ignored them

Agreed. I will get it working this way.

Reply:
@caleb-johnson , I just remembered that my thought process behind this was that if the user gave us a (contrived) circuit with, say, just one 3-qubit gate and asked us to cut it, it should through an error. Perhaps we should go with what you're advocating and say that only 2-qubit gates can be cut somewhere in the documentation?

Reply:
Yes, that's exactly the case I created to test this. I made a 7-qubit circuit with a ladder of CCXGates. I would expect the optimizer to either tell me it couldn't find a solution or to return me the solution, regardless of the input circuit. How hard would it be to make it behave like that?

Reply:
Do you mean how hard it would be to get it to cut arbitrary qubit gates? There was some machinery there for cutting multi-qubit gates at one point and we took it out (after some discussion) because we said that cutting 2 qubit gates would be the most immediate use case. Another motivation for doing this was that the transpiler converts all input gates into single and two qubit gates.

Reply:
Having looked through the changes I've made in other places in the code (since we discussed removing support for gates with 3 or more inputs), the only consistent thing to do would be to throw an error when a gate with more than 2 qubits is encountered in an input circuit. Changing this would involve adding back some the machinery I talked about above, even if the goal is just apply that gate and ignore cutting it. We can discuss whether than needs to be done but it would be a little non-trivial.

Reply:
I think it's OK for the MVP to not accept inputs containing ``N>2``-qubit gates. I'll catch that error in `find_cuts` and tell users to decompose their `N`-qubit gates into single and two-qubit operations